File: eval/solutions/go/06-errors-interfaces/main_test.go
package main

import "testing"

func TestBatterySensorValid(t *testing.T) {
	r := NewRobot("R2D2")
	r.Battery = 75
	s := BatterySensor{robot: r}
	val, err := s.Read()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if val != 75 {
		t.Errorf("got %d, want 75", val)
	}
}

func TestBatterySensorInvalidNegative(t *testing.T) {
	r := NewRobot("R2D2")
	r.Battery = -1
	s := BatterySensor{robot: r}
	_, err := s.Read()
	if err == nil {
		t.Error("expected error for battery -1, got nil")
	}
}

func TestBatterySensorInvalidOver100(t *testing.T) {
	r := NewRobot("R2D2")
	r.Battery = 101
	s := BatterySensor{robot: r}
	_, err := s.Read()
	if err == nil {
		t.Error("expected error for battery 101, got nil")
	}
}

func TestReadSensor(t *testing.T) {
	r := NewRobot("R2D2")
	r.Battery = 80
	s := BatterySensor{robot: r}
	val, err := ReadSensor(s)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if val != 80 {
		t.Errorf("got %d, want 80", val)
	}
}

func TestSensorInterface(t *testing.T) {
	r := NewRobot("R2D2")
	r.Battery = 50
	var _ Sensor = BatterySensor{robot: r}
}

File: rust/06-errors-traits/tests/main_test.rs
use lesson_06_errors_traits::{Robot, BatterySensor, Sensor, read_sensor};

#[test]
fn test_battery_sensor_valid() {
    let r = Robot::new("R2D2");
    let s = BatterySensor { robot: &r };
    assert_eq!(s.read(), Ok(100));
}

#[test]
fn test_battery_sensor_invalid_negative() {
    let mut r = Robot::new("R2D2");
    r.battery = -1;
    let s = BatterySensor { robot: &r };
    assert!(s.read().is_err());
}

#[test]
fn test_battery_sensor_invalid_over_100() {
    let mut r = Robot::new("R2D2");
    r.battery = 101;
    let s = BatterySensor { robot: &r };
    assert!(s.read().is_err());
}

#[test]
fn test_read_sensor() {
    let mut r = Robot::new("R2D2");
    r.battery = 80;
    let s = BatterySensor { robot: &r };
    assert_eq!(read_sensor(&s), Ok(80));
}
